src/lib/device/virtual_dev.cpp

```diff
@@ -426,54 +167,16 @@ BlkAllocStatus VirtualDev::alloc_blk(uint32_t nblks, const blk_alloc_hints& hint
 BlkAllocStatus VirtualDev::do_alloc_blk(blk_count_t nblks, const blk_alloc_hints& hints,
                                         std::vector< BlkId >& out_blkid) {
     try {
-        PhysicalDevChunk* preferred_chunk = nullptr;
-        auto* stream_info = (stream_info_t*)(hints.stream_info);
-        uint32_t try_streams = 0;
-        if (stream_info && (stream_info->num_streams != 0)) { try_streams = stream_info->num_streams; }
-
-        BlkAllocStatus status = BlkAllocStatus::FAILED;
-
-        while (try_streams != 0) {
-            preferred_chunk = reinterpret_cast< PhysicalDevChunk* >(stream_info->chunk_list[stream_info->stream_cur]);
-            if (preferred_chunk) {
-                // try to allocate from the preferred chunk
-                status = alloc_blk_from_chunk(nblks, hints, out_blkid, preferred_chunk);
-                if (status == BlkAllocStatus::SUCCESS) { return status; };
-            }
-            stream_info->stream_cur = (stream_info->stream_cur + 1) % stream_info->num_streams;
-            --try_streams;
-        }
-
-        if (stream_info) {
-            // try to allocate from the default chunk
-            status = alloc_blk_from_chunk(nblks, hints, out_blkid, m_default_chunk);
-            if (status == BlkAllocStatus::SUCCESS) {
-                COUNTER_INCREMENT(m_metrics, default_chunk_allocation_cnt, 1);
-                return status;
-            }
-        }
-
-        if (m_pdev_group == PhysicalDevGroup::DATA) { COUNTER_INCREMENT(m_metrics, random_chunk_allocation_cnt, 1); }
+        Chunk* first_failed_chunk = nullptr;
 
-        // try to allocate from the other chunks now
-        // First select a device to allocate from
-        uint32_t chunk_num, start_chunk_num;
-        uint32_t dev_ind{0};
-        dev_ind = (hints.dev_id_hint == INVALID_DEV_ID) ? m_selector->select(hints) : uint32_cast(hints.dev_id_hint);
-
-        // Pick a physical chunk based on physDevId.
-        // TODO: Right now there is only one primary chunk per device in a virtualdev. Need to support multiple
-        // chunks. In that case just using physDevId as chunk number is not right strategy.
-        uint32_t start_dev_ind = dev_ind;
+        // First select a chunk to allocate it from
+        BlkAllocStatus status;
+        Chunk* chunk;
         do {
-            for (auto& chunk : m_primary_pdev_chunks_list[dev_ind].chunks_in_pdev) {
-                status = alloc_blk_from_chunk(nblks, hints, out_blkid, chunk);
-                if (status == BlkAllocStatus::SUCCESS || !hints.can_look_for_other_chunk) { break; }
-            }
-
+            chunk = m_chunk_selector->select(hints);
```

Comment:
1. if return chunk is nullptr, need break;
2. first_failed_chunk is  never set.

Reply:
Fixed that